rids/rules.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 Jigsaw Operations LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Generalizes the various formats of IOCs into Rule and RuleSet.
+
+Rule is a value type that represents evaluation on a single rule.  RuleSet
+instances represent a set of independent and overlapping Rule instances.
+
+RuleSet should typically be constructed from one of the flavors of parsers
+in `rids.iocs` or merged from combining a pair of RuleSet instances.  Its
+
+Evaluation of the RuleSet is done with the ProcessHandshake(...) and
+ProcessEndpoint(...) methods which take an observation and return a list
+of zero or more Event instances with their matching rules in the RuleSet.
+"""
+
+
+import collections
+
+
+class RuleSet:
+    __slots__ = ('ip_address_rules', )
+
+    def __init__(self):
+      self.ip_address_rules = collections.defaultdict(list)
+
+    def MergeRuleset(self, other):
+      """Simple implementation of a merging of rule sets."""
+      for key, value in other.ip_address_rules.items():
+        self.ip_address_rules[key].extend(value)
+
+    def AddRule(self, rule):
+      """Simple implementation of adding a single rule to this rule set."""
+      if hasattr(rule, 'matches_ip'):
+        self.ip_address_rules[str(rule.matches_ip)].append(rule)
+
+    def ProcessHandshake(self, observation):
```

Comment:
I don't think this method is doing much. Perhaps we leave this class just for IP matching for now.

Reply:
It produces the Event representation so that main() can log it or however we process it in the future.  It just so happens that with the current capture definition for handshakes it only produces the suspicious ones so these always turn into Event.  I commented a TODO that perhaps we could put the preprocessing here, though that would mean seeing client_hello data twice and would make RuleSet stateful, which I don't think is great (it would complicate the update for the ruleset when IOC updates are fetched periodically).

I could leave the client/server hello matching in the network scanner and put just the allow-list handling here, which does align nicely with the concept of RuleSet, but do you mind if I do that in a future PR?  It would justify leaving this simplistic ProcessHandshake here for now.

Reply:
it now does the allow-list checking here